# Лабораторная работа №1 (Алгоритм Евклида)
Выполнил: студент группы ИДМ-24-07 Туркин Александр

#### 1. Описание алгоритма Евклида
Алгоритм Евклида — это метод нахождения наибольшего общего делителя (НОД) двух целых чисел. Он основан на следующем принципе: НОД двух чисел a и b равен НОД числа b и остатка от деления a на b, то есть:

$НОД(a, b) = НОД(b, a \mod b) $

Алгоритм продолжается до тех пор, пока одно из чисел не станет равным нулю. В этом случае другое число является НОД.

#### 2. Структура представления данных в программе
В программе используются следующие структуры данных:
- **Целые числа** для хранения входных значений (a и b).
- **Массивы** для хранения значений, если необходимо обрабатывать несколько пар чисел.
- **Строки** для работы с файлами и вводом/выводом данных.

#### 3. Основные функции программы
Программа включает следующие функции:

1. **`int gcd(int a, int b)`**:
   - **Назначение**: Вычисляет НОД двух целых чисел.
   - **Входные параметры**: два целых числа (a, b).
   - **Выходные параметры**: НОД (целое число).

In [1]:
def gcd(a, b):
    """Функция для вычисления наибольшего общего делителя (НОД)"""
    while b != 0:
        a, b = b, a % b
    return a

2. **`void readFromFile(const char* filename)`**:
   - **Назначение**: Читает пары чисел из файла.
   - **Входные параметры**: имя файла (строка).
   - **Выходные параметры**: нет (результаты сохраняются в глобальных переменных или передаются в другие функции).

In [2]:
def read_from_file(filename):
    """Чтение пар чисел из файла"""
    pairs = []
    with open(filename, 'r') as file:
        for line in file:
            a, b = map(int, line.split())
            pairs.append((a, b))
    return pairs


3. **`void readFromInput()`**:
   - **Назначение**: Считывает пары чисел с клавиатуры.
   - **Входные параметры**: нет.
   - **Выходные параметры**: нет (результаты сохраняются в глобальных переменных или передаются в другие функции).

In [3]:
def read_from_input():
    """Считывание пар чисел с клавиатуры"""
    pairs = []
    while True:
        user_input = input("Введите два числа через пробел (или 'exit' для выхода): ")
        if user_input.lower() == 'exit':
            break
        try:
            a, b = map(int, user_input.split())
            pairs.append((a, b))
        except ValueError:
            print("Ошибка: введите два целых числа.")
    return pairs


4. **`void writeToFile(const char* filename, int results)`**:
   - **Назначение**: Записывает результат в файл.
   - **Входные параметры**: имя файла (строка), результат (массив с ответами).
   - **Выходные параметры**: нет.

In [4]:
def write_to_file(filename, results):
    """Запись результатов в файл"""
    with open(filename, 'w') as file:
        for a, b, result in results:
            file.write(f"НОД({a}, {b}) = {result}\n")

#### 4. Функция main и проверка работы программы.

Для проверки напишем скрипт для создания файла с рандомными числами. Файл называется random_pairs.txt в котором 100 пар рандомных чисел от 0 до 1000000.

In [6]:
import random

# Указываем имя файла
filename = 'random_pairs.txt'

# Открываем файл для записи
with open(filename, 'w') as file:
    for _ in range(100):
        # Генерируем случайные числа
        num1 = random.randint(0, 1000000)
        num2 = random.randint(0, 1000000)
        # Записываем пару чисел в файл
        file.write(f"{num1} {num2}\n")

print(f"Файл '{filename}' успешно создан с 100 парами случайных чисел.")

Файл 'random_pairs.txt' успешно создан с 100 парами случайных чисел.


Функция main() программы.

In [7]:
def main():
    choice = input("Выберите способ ввода (1 - файл, 2 - клавиатура): ")
    pairs = []

    if choice == '1':
        filename = input("Введите имя файла: ")
        pairs = read_from_file(filename)
    elif choice == '2':
        pairs = read_from_input()
    else:
        print("Неверный выбор.")
        return

    results = []
    for a, b in pairs:
        result = gcd(a, b)
        results.append((a, b, result))
        print(f"НОД({a}, {b}) = {result}")

    output_filename = input("Введите имя файла для записи результатов: ")
    write_to_file(output_filename, results)

if __name__ == "__main__":
    main()

НОД(478301, 139377) = 1
НОД(690000, 46586) = 2
НОД(778898, 66738) = 2
НОД(894923, 171008) = 1
НОД(905574, 688451) = 1
НОД(530621, 131457) = 1
НОД(257186, 979528) = 2
НОД(472273, 266747) = 1
НОД(233471, 134901) = 1
НОД(450580, 534324) = 4
НОД(326164, 915613) = 1
НОД(69044, 217084) = 4
НОД(928275, 901394) = 1
НОД(958079, 755202) = 1
НОД(94049, 647508) = 1
НОД(401251, 49899) = 1
НОД(610245, 411341) = 1
НОД(685698, 613252) = 2
НОД(296511, 123006) = 3
НОД(238533, 979103) = 1
НОД(297655, 794613) = 1
НОД(940061, 244804) = 1
НОД(568170, 231085) = 5
НОД(844597, 649734) = 1
НОД(122623, 165241) = 1
НОД(110938, 222495) = 1
НОД(283399, 768818) = 1
НОД(520107, 789851) = 1
НОД(917656, 303991) = 1
НОД(209932, 290252) = 4
НОД(702213, 994224) = 3
НОД(824205, 28111) = 1
НОД(915873, 560533) = 1
НОД(12382, 516740) = 2
НОД(571878, 678734) = 2
НОД(909608, 840797) = 1
НОД(175483, 848876) = 7
НОД(916857, 627374) = 1
НОД(591874, 645492) = 2
НОД(472903, 297897) = 1
НОД(233056, 95540) = 4
НОД(604077, 456272) = 1


Результат так же был записан в файл result.txt


#### 5. Контрольные вопросы

1. **Определение простого числа**:
   Простое число — это натуральное число больше 1, которое имеет ровно два различных делителя: 1 и само число.

2. **Определение составного числа**:
   Составное число — это натуральное число больше 1, которое имеет более двух делителей.

3. **Алгоритм Евклида**:
   1. Вводим два числа a и b.
   2. Пока b не равно 0:
      - Вычисляем остаток от деления a на b (r = a mod b).
      - Присваиваем a значение b, а b значение r.
   3. Когда b равно 0, a является НОД.

4. **Определение наибольшего общего делителя (НОД)**:
   НОД двух целых чисел — это наибольшее число, на которое оба числа делятся без остатка.

5. **Теорема о делении двух целых чисел**:
   Для любых двух целых чисел a и b (где b ≠ 0) существуют такие целые числа $q$ и $r$, что:
   $ a = b \cdot q + r $
, где $0 ≤ r < |b|$. Здесь $q$ — это частное, а $r$ — остаток от деления a на b.

#### 6. Приложение
|random_pairs.txt|result.txt|
|:--------------:|:--------:|
|478301 139377 | НОД(478301, 139377) = 1
690000 46586 | НОД(690000, 46586) = 2
778898 66738 | НОД(778898, 66738) = 2
894923 171008 | НОД(894923, 171008) = 1
905574 688451 | НОД(905574, 688451) = 1
530621 131457 | НОД(530621, 131457) = 1
257186 979528 | НОД(257186, 979528) = 2
472273 266747 | НОД(472273, 266747) = 1
233471 134901 | НОД(233471, 134901) = 1
450580 534324 | НОД(450580, 534324) = 4
326164 915613 | НОД(326164, 915613) = 1
69044 217084 | НОД(69044, 217084) = 4
928275 901394 | НОД(928275, 901394) = 1
958079 755202 | НОД(958079, 755202) = 1
94049 647508 | НОД(94049, 647508) = 1
401251 49899 | НОД(401251, 49899) = 1
610245 411341 | НОД(610245, 411341) = 1
685698 613252 | НОД(685698, 613252) = 2
296511 123006 | НОД(296511, 123006) = 3
238533 979103 | НОД(238533, 979103) = 1
297655 794613 | НОД(297655, 794613) = 1
940061 244804 | НОД(940061, 244804) = 1
568170 231085 | НОД(568170, 231085) = 5
844597 649734 | НОД(844597, 649734) = 1
122623 165241 | НОД(122623, 165241) = 1
110938 222495 | НОД(110938, 222495) = 1
283399 768818 | НОД(283399, 768818) = 1
520107 789851 | НОД(520107, 789851) = 1
917656 303991 | НОД(917656, 303991) = 1
209932 290252 | НОД(209932, 290252) = 4
702213 994224 | НОД(702213, 994224) = 3
824205 28111 | НОД(824205, 28111) = 1
915873 560533 | НОД(915873, 560533) = 1
12382 516740 | НОД(12382, 516740) = 2
571878 678734 | НОД(571878, 678734) = 2
909608 840797 | НОД(909608, 840797) = 1
175483 848876 | НОД(175483, 848876) = 7
916857 627374 | НОД(916857, 627374) = 1
591874 645492 | НОД(591874, 645492) = 2
472903 297897 | НОД(472903, 297897) = 1
233056 95540 | НОД(233056, 95540) = 4
604077 456272 | НОД(604077, 456272) = 1
663403 866393 | НОД(663403, 866393) = 1
453395 341495 | НОД(453395, 341495) = 5
820341 978698 | НОД(820341, 978698) = 1
109096 443230 | НОД(109096, 443230) = 2
328185 843586 | НОД(328185, 843586) = 1
33002 377941 | НОД(33002, 377941) = 1
642013 164467 | НОД(642013, 164467) = 1
999971 969762 | НОД(999971, 969762) = 1
490106 706947 | НОД(490106, 706947) = 1
127476 450012 | НОД(127476, 450012) = 12
554902 150729 | НОД(554902, 150729) = 1
850463 549980 | НОД(850463, 549980) = 1
708385 2688 | НОД(708385, 2688) = 1
446199 886113 | НОД(446199, 886113) = 3
378575 304321 | НОД(378575, 304321) = 1
970727 182052 | НОД(970727, 182052) = 1
278613 306891 | НОД(278613, 306891) = 9
323476 180982 | НОД(323476, 180982) = 34
989869 307433 | НОД(989869, 307433) = 1
200442 441245 | НОД(200442, 441245) = 1
315989 944834 | НОД(315989, 944834) = 1
629148 89335 | НОД(629148, 89335) = 1
221269 930577 | НОД(221269, 930577) = 1
449570 157983 | НОД(449570, 157983) = 1
355840 845109 | НОД(355840, 845109) = 1
480023 80591 | НОД(480023, 80591) = 1
275973 733583 | НОД(275973, 733583) = 67
527292 937700 | НОД(527292, 937700) = 4
4074 669749 | НОД(4074, 669749) = 1
612023 145649 | НОД(612023, 145649) = 1
758785 593337 | НОД(758785, 593337) = 1
343417 348351 | НОД(343417, 348351) = 1
558864 108724 | НОД(558864, 108724) = 4
263420 939507 | НОД(263420, 939507) = 1
155191 524952 | НОД(155191, 524952) = 1
530627 538358 | НОД(530627, 538358) = 1
658618 561074 | НОД(658618, 561074) = 2
171956 915886 | НОД(171956, 915886) = 2
297234 527097 | НОД(297234, 527097) = 3
393890 359389 | НОД(393890, 359389) = 1
984941 438595 | НОД(984941, 438595) = 1
733762 199850 | НОД(733762, 199850) = 2
938612 629238 | НОД(938612, 629238) = 2
392633 666811 | НОД(392633, 666811) = 1
928707 90621 | НОД(928707, 90621) = 3
338125 151741 | НОД(338125, 151741) = 1
95931 44146 | НОД(95931, 44146) = 1
404850 329569 | НОД(404850, 329569) = 1
424704 909356 | НОД(424704, 909356) = 28
599624 90229 | НОД(599624, 90229) = 1
363132 411550 | НОД(363132, 411550) = 2
135404 56004 | НОД(135404, 56004) = 4
267379 511030 | НОД(267379, 511030) = 1
625417 888089 | НОД(625417, 888089) = 1
593008 932946 | НОД(593008, 932946) = 2
210649 281748 | НОД(210649, 281748) = 1
606166 181515 | НОД(606166, 181515) = 1
797638 817681 | НОД(797638, 817681) = 1|